OutOfSchool/OutOfSchool.WebApi/Services/Gcp/GcpImagesStorageSynchronizationService.cs

```diff
@@ -0,0 +1,138 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Apis.Storage.v1.Data;
+using Google.Cloud.Storage.V1;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using OutOfSchool.Services.Models;
+using OutOfSchool.Services.Models.Images;
+using OutOfSchool.Services.Repository;
+using OutOfSchool.Services.Repository.Files;
+using OutOfSchool.WebApi.Models;
+using Object = Google.Apis.Storage.v1.Data.Object;
+
+namespace OutOfSchool.WebApi.Services.Gcp
+{
+    /// <summary>
+    /// Represents a gcp image files sync storage. It's used to synchronize gcp files with a database.
+    /// </summary>
+    public class GcpImagesStorageSynchronizationService : IGcpStorageSynchronizationService
+    {
+        private const string ListObjectOptionsFields = "items(name,timeCreated),nextPageToken";
+
+        // TODO: must be changed to approximately 500 after testing
+        private const int ListObjectOptionsPageSize = 50;
+
+        private readonly ILogger<GcpImagesStorageSynchronizationService> logger;
+        private readonly IImageFilesStorage imagesStorage;
+        private readonly IGcpImagesSyncDataRepository gcpImagesSyncDataRepository;
+
+        public GcpImagesStorageSynchronizationService(
+            ILogger<GcpImagesStorageSynchronizationService> logger,
+            IImageFilesStorage imagesStorage,
+            IGcpImagesSyncDataRepository gcpImagesSyncDataRepository)
+        {
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            this.imagesStorage = imagesStorage ?? throw new ArgumentNullException(nameof(imagesStorage));
+            this.gcpImagesSyncDataRepository = gcpImagesSyncDataRepository ?? throw new ArgumentNullException(nameof(gcpImagesSyncDataRepository));
+
+            GetAllSyncFunctions = new List<Func<IEnumerable<string>, Task<List<string>>>>
+            {
+                // Entity images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderImagesIds,
+
+                // Entity cover images
+                gcpImagesSyncDataRepository.GetIntersectWorkshopCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectTeacherCoverImagesIds,
+                gcpImagesSyncDataRepository.GetIntersectProviderCoverImagesIds,
+            };
+        }
+
+        private List<Func<IEnumerable<string>, Task<List<string>>>> GetAllSyncFunctions { get; }
+
+        /// <inheritdoc/>
+        public async Task SynchronizeAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                logger.LogDebug("Gcp storage synchronization was started");
+
+                // gcp returns data by local time
+                var dateTime = DateTime.Now.AddHours(-1);
```

Comment:
-1 is needed to give all started processes oportunity to finish. Whenever we start this job and some requests are made by users, we can exclude the situation we create files in gcp but not save info about them in our main database because of 1 hour gap.
I've already tried to use UTC but gcp returns data using local server TZ so I guess we can use it? The problem can be if TZ is changed in the middle of synchronization that is unbelievable. If TZ is changed in server then another TZ will be chosen by gcp.
Maybe I've missed something in my thoughts.

Reply:
You've stepped into the usual problem with TZ with DST (day saving time). 
An issue might be when TZ is not changed, but your DST is switched on/off. I.e. in Ukraine you idea with -1 hour will not work twice a year at 3 AM last Sunday on March and October.

It's minor, but it's better to avoid it using UTC (and ask gcp about UTC time, if it is possible)

Reply:
We can continue this discussion about `DateTime` and .NET for days or weeks or... ))
I will try to rephrase my question: 
1. why not -5 ? or -3.762056 ?
2. Also 'unbelievable' for me literally means: we know there is a bug which will bubble up once per few years and nobody knows/remembers what and why, and everyone will continue to suffer )

Also, just imagine how much efforts we will have to put here to identify all places which should be updated in case if GCP changed their free client library or data format or timing format ?
Better option, as for me, is to have a single place for converting input data to format that we will use in the whole application, so the only single place to update in case of any changes.
Does that make sense ?

Reply:
1. **Agree with bouth of you, thanks.** I've just converted `DateTime` that I get from gcp to utc where I compare and filtering with UtcNow.
2. @h4wk13 About -5, -3.762056 I moved it into constants but I also want to determine what I mean. This isn't about time zone evaluating. It's about the situation when some user starts operation for example 'creating workshop' then some images will be uploaded into gcp storage but there are also few long time tasks before changes are applied to our database so without 'operation gap' can be moments when we delete files from gcp by synchronizer because they are missing in our db then after some time we commit changes into our db. So we have ids in our db which were deleted by synchronizer.
3. That's because we decided not to have a distributed transaction between our db and gcp **at least for now** and synchronizer can be triggered when users do their jobs (we don't close ports or anything else while syncronization).
Also that's why the trigger must be fired at night.